server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,533 @@
+        PostParsePhase postParsePhase = lookup.buildPostParsePhase(parseContext);
+        if (postParsePhase == null) {
+            return;
+        }
+        postParsePhase.execute();
+    }
+
+    PostParsePhase(
+        Map<String, PostParseExecutor> postParseExecutors,
+        Function<String, MappedFieldType> fieldTypeLookup,
+        ParseContext pc) {
+        LazyDocumentReader reader = new LazyDocumentReader(
+            pc.rootDoc(),
+            pc.sourceToParse().source(),
+            postParseExecutors.keySet());
+        this.context = new PostParseContext(fieldTypeLookup, pc, reader.getContext());
+        postParseExecutors.forEach((k, c) -> fieldExecutors.put(k, new OneTimeFieldExecutor(c)));
+    }
+
+    void execute() throws IOException {
+        for (OneTimeFieldExecutor executor : fieldExecutors.values()) {
+            executor.execute();
+        }
+    }
+
+    // FieldExecutors can be called both by executePostParse() and from the lazy reader,
+    // so to ensure that we don't run field scripts multiple times we guard them with
+    // this one-time executor class
+    private class OneTimeFieldExecutor {
+
+        PostParseExecutor executor;
+        boolean executed = false;
+
+        OneTimeFieldExecutor(PostParseExecutor executor) {
+            this.executor = executor;
+        }
+
+        void execute() throws IOException {
+            if (executed == false) {
+                try {
+                    executor.execute(context);
+                } catch (Exception e) {
+                    executor.onError(context, e);
+                }
+                executed = true;
+            }
+        }
+    }
+
+    private class LazyDocumentReader extends LeafReader {
+
+        private final ParseContext.Document document;
+        private final BytesReference sourceBytes;
+        private final Set<String> calculatedFields;
+        private final Set<String> fieldPath = new LinkedHashSet<>();
+
+        private LazyDocumentReader(ParseContext.Document document, BytesReference sourceBytes, Set<String> calculatedFields) {
+            this.document = document;
+            this.sourceBytes = sourceBytes;
+            this.calculatedFields = calculatedFields;
+        }
+
+        private void checkField(String field) throws IOException {
+            if (calculatedFields.contains(field)) {
+                // this means that a mapper script is referring to another calculated field;
+                // in which case we need to execute that field first. We also check for loops here
+                if (fieldPath.add(field) == false) {
+                    throw new IllegalStateException("Loop in field resolution detected: " + String.join("->", fieldPath) + "->" + field);
+                }
+                assert fieldExecutors.containsKey(field);
+                fieldExecutors.get(field).execute();
+                calculatedFields.remove(field);
```

Comment:
I am missing why we remove from this set, is it so we execute the script only once?

Reply:
Yes - once it has been executed then its values are in the document and so will get picked up by the lazy reader normally, so we don't need to do this dance if it is referred to a second time.

Reply:
ok I think I get confused because I see executing the script once being a concern in two different places, one being the one time executor, and the other one being this set.

Reply:
Yes, it's not strictly necessary to remove it here but it means we can skip this whole branch once the field has been intercepted once.  I can remove it if you think it confuses things too much though.

Reply:
leave it for now, I will bump into it again in my next review and see